### PR TITLE
Update MaliciousOAuthApp_O365AttackToolkit.yaml

### PR DESCRIPTION
**Pull request by treyperrone**
Required items, please complete
   
   Change(s):
   - Create dynamic array for older o365 default scope
   - Create dynamic array for newer o365 default scope
   - has_all for both arrays to detect either usage of the tool. Best guess average attacker uses the second "new" default scope since its out-of-the-box

   Reason for Change(s):
   - Alert not firing with updates made to O365 toolkit November 2020
   - Current rule uses an older defaults for `Consent.Full` that was deprecated from the file: [https://github.com/mdsecactivebreach/o365-attack-toolkit/blob/2231565f576e2d655cb9444ffed7309374c454a5/static/index.html#L29](https://github.com/mdsecactivebreach/o365-attack-toolkit/blob/2231565f576e2d655cb9444ffed7309374c454a5/static/index.html#L29)
   - The NEW defaults generated in Nov 2020 are here: [https://github.com/mdsecactivebreach/o365-attack-toolkit/blob/master/template.conf ](https://github.com/mdsecactivebreach/o365-attack-toolkit/blob/master/template.conf )
   - old values to match on: `"offline_access","contacts.read","user.read","mail.read","notes.read.all","mailboxsettings.readwrite","Files.ReadWrite.All"`
   - NEW values to match on: `"offline_access contacts.read user.read mail.read mail.send files.readWrite.all files.read files.read.all openid profile"`
   - The **current detection** relies on finding `mailboxsettings` in LN32 [https://github.com/Azure/Azure-Sentinel/blob/c6dce9c3aa4d4b4d02423ac4eb5a6b677a39e432/Solutions/Azure%20Active%20Directory/Analytic%20Rules/MaliciousOAuthApp_O365AttackToolkit.yaml#L32](https://github.com/Azure/Azure-Sentinel/blob/c6dce9c3aa4d4b4d02423ac4eb5a6b677a39e432/Solutions/Azure%20Active%20Directory/Analytic%20Rules/MaliciousOAuthApp_O365AttackToolkit.yaml#L32) which means anyone that downloads the current O365 attack toolkit and runs it with the defaults settings will NOT trigger the alert.

   Version Updated:
   - Yes

   Testing Completed:
   - Yes

   Checked that the validations are passing and have addressed any issues that are present:
   - Yes

### CODE REVIEW

**Comment by v-mchatla** [comment]
Hi @devikamehra, @vakohl,
Can you please review and share your feedback.
Thanks

File: Solutions/Azure Active Directory/Analytic Rules/MaliciousOAuthApp_O365AttackToolkit.yaml
Comment (devikamehra) [reply Solutions/Azure Active Directory/Analytic Rules/MaliciousOAuthApp_O365AttackToolkit.yaml:28]: It would be great if you can share some samples to help make testing easier.

Comment (treyperrone) [reply Solutions/Azure Active Directory/Analytic Rules/MaliciousOAuthApp_O365AttackToolkit.yaml:28]: I can try and grab some samples from testing, but I am not really on the red-team side setting up the o365 attack side and just modifying what another person did with the o365 toolkit they downloaded directly off github repo.

If Microsoft has someone more familiar with the process or someone from DART that can assist, that may be useful.

I think the main issue is; if we search for ALL scopes (as the current rule was written) we are detecting an attack from someone who runs the toolkit "_out-of-the-box._" A smart or advanced actor could simply view this same source code we are editing, since public, or stand up their own Azure account and test and realize that scoping to just a few permissions they need for their attack would not trigger an alert in Sentinel and may go unnoticed?

Comment (devikamehra) [reply Solutions/Azure Active Directory/Analytic Rules/MaliciousOAuthApp_O365AttackToolkit.yaml:28]: @treyperrone Thank you so much for clarifying. My intention here is to understand what is the minimum scope required to perform this activity. This would help us in performing "has_all" operation. Otherwise, we would end up missing a new of these instances.
I will try to see if we have someone internally who can help us here but if you have the required information, that would be great,
Also, please send the sample logs to demehra@microsoft.com

Comment (treyperrone) [reply Solutions/Azure Active Directory/Analytic Rules/MaliciousOAuthApp_O365AttackToolkit.yaml:28]: - `has_all` does indeed need all items to fire the alert
- `has_any` may be more appropriate
- we also need to modify [https://github.com/Azure/Azure-Sentinel/blob/cff4ebd4b3efe69331d385d3bf5fa9a88f0c1d35/Solutions/Azure%20Active%20Directory/Analytic%20Rules/MaliciousOAuthApp_O365AttackToolkit.yaml#L34](https://github.com/Azure/Azure-Sentinel/blob/cff4ebd4b3efe69331d385d3bf5fa9a88f0c1d35/Solutions/Azure%20Active%20Directory/Analytic%20Rules/MaliciousOAuthApp_O365AttackToolkit.yaml#L34) and perhaps fully remove `| where TargetResources has "<some value>"` as it may not serve much of a purpose with searching for all values.

**I still dont know WHAT if all values should be the correct ones here.** With the simple test of only `mail.send` someone can only send mail as that user, but that alone can be damaging if its a malicious attacker sending mail masquerading as an actual user. 

more verbose output of a query test with ONLY `mail.send` 
[query_data_full-output.csv](https://github.com/Azure/Azure-Sentinel/files/10843437/query_data_full-output.csv)

screenshot with `has_any` only (no match)
Screenshot 2023-02-27 at 2 37 20 PM


screenshot with `has_any` only (no match)
Screenshot 2023-02-27 at 2 38 13 PM


screenshot with single dynamic array all values, `has_any` detect, removal of `| where TargetResources has "value"`
Screenshot 2023-02-27 at 2 56 08 PM

log from this last event above: 
[query_data_single_array.csv](https://github.com/Azure/Azure-Sentinel/files/10843506/query_data_single_array.csv)

Comment (treyperrone) [reply Solutions/Azure Active Directory/Analytic Rules/MaliciousOAuthApp_O365AttackToolkit.yaml]: @devikamehra  see comments above and new updates to the KQL.

Comment (devikamehra) [reply Solutions/Azure Active Directory/Analytic Rules/MaliciousOAuthApp_O365AttackToolkit.yaml]: @treyperrone Just to ensure we are on the same page, you are referring to https://github.com/mdsecactivebreach/o365-attack-toolkit for simulating the attack in this scenario?
If this is true, we should do the following:
1. Remove the TargetResource comparison line as this is a repetition.
2. Move back to the approach you took earlier with two lists. In the attack2 scenario, remove the file read related permissions as these are already inherited with ReadWrite. And now use has_all with respective lists. 

Also, the record that you sent doesn't seems to be the result of the attack documented in the link. If we have has_any operator with such wide scoped grants, the rule will have high number of false positives.

Please let me know in case more information is required from my end.

Comment (treyperrone) [reply Solutions/Azure Active Directory/Analytic Rules/MaliciousOAuthApp_O365AttackToolkit.yaml]: - Yes, we used the attack-toolkit
- the pentester scoped the permissions to only `mail.send` for the example
 
 if we set it back to looking for two `has_all` variants, all we are detecting is an attacker that is not smart enough to work around the sentinel detection we have publicly available on github, so the detection is fair less useful against advanced or persistent attackers

Comment (devikamehra) [reply Solutions/Azure Active Directory/Analytic Rules/MaliciousOAuthApp_O365AttackToolkit.yaml]: @treyperrone Understanding the concern here, we can modify the query in the following way

let detectionTime = 1d;  let joinLookback = 14d;  let threshold = 5;  let o365_attack_regex = "contacts.read|user.read|mail.read|notes.read.all|mailboxsettings.readwrite|Files.ReadWrite.All|mail.send|files.read|files.read.all";  let o365_attack = dynamic(["contacts.read","user.read","mail.read","notes.read.all","mailboxsettings.readwrite","Files.ReadWrite.All","mail.send","files.read","files.read.all"]);  let apps = dynamic(['5523c0ac-586d-491e-bc73-fc18f9ed2c28','3190d40e-82d9-44ee-af78-4ce2aafd4f15','0771c6f9-71c7-46d4-87a9-3b57443ee250','a644e37f-d354-440f-843a-1e1ee7dfc4ef','21fbe0e7-6e83-4fb0-a56f-92a67c64e37a','30966914-6789-4d1c-9e2b-76a81a5e2a54','2a6082d8-9f0a-455c-a26b-7fce81edefb8','adc8d359-f3ef-4e0a-b170-2bcf5a022728']);  AuditLogs  | where TimeGenerated > ago(detectionTime)  | where LoggedByService =~ "Core Directory"  | where Category =~ "ApplicationManagement"  | where OperationName =~ "Consent to application"  | extend AppDisplayName = TargetResources.[0].displayName  | extend AppClientId = tolower(TargetResources.[0].id)  | where AppClientId !in (apps)  | extend ConsentFull = TargetResources[0].modifiedProperties[4].newValue  | parse ConsentFull with * "ConsentType: " GrantConsentType ", Scope: " GrantScope1 ", CreatedDateTime" * "]" *  | where GrantConsentType != "AllPrincipals" // NOTE: we are ignoring if OAuth application was granted to all users via an admin - but admin due diligence should be audited occasionally  | where ConsentFull has_any (o365_attack)  | extend GrantScopeCount = countof(tolower(GrantScope1), o365_attack_regex, 'regex')  | where GrantScopeCount > threshold

If we perform check on has_any, this alert can become noisy and thus would create a huge list of false positive. Thus, adding a threshold here which can prevent this from happening. If you feel the version you suggested would work better in your environment, please tune the threshold parameter to 1 and it should solve the problem.

Let me know if you have any more questions.

**Comment by v-mchatla** [comment]
Hi @treyperrone,
Thanks for making the changes.
@devikamehra - Can you please review the changes and suggest author if there is any other changes that need to be made.
Thanks

**Comment by devikamehra** [comment]
@v-mchatla I have already provided updates on the changes the author has made.
Will wait for him to respond on the changes suggested.

File: Solutions/Azure Active Directory/Analytic Rules/MaliciousOAuthApp_O365AttackToolkit.yaml
Comment (treyperrone) [reply Solutions/Azure Active Directory/Analytic Rules/MaliciousOAuthApp_O365AttackToolkit.yaml]: Sorry for the delay, trying to get more input from other security team members in our organization with more diverse experience and opinions :-)

The threshold suggestion is an excellent one that sets the bar for the average Sentinel user and lets everyone tweak it with ease!

Migrating the external watchlist inline also is good, because the rules are versioned but a change to the external CSV wouldn't be tracked. I would suggest we a minor name tweak to `trusted_apps`, a list format with comments/names for easier end user experience (the CSV has the names for clarity). This list format also puts the closing parenthesis on a newline so easier to comment values out and keep the dynamic array able to pass a NULL value (minor thing, but helpful for newer KQL users).

so in full the potential full playbook could look like below. If these seems good happy to update the patch-2 branch again

```
// let detectionTime = 1d;
// let joinLookback = 14d;
let threshold = 5;
let o365_attack_regex = "contacts.read|user.read|mail.read|notes.read.all|mailboxsettings.readwrite|Files.ReadWrite.All|mail.send|files.read|files.read.all";
let o365_attack = dynamic(["contacts.read", "user.read", "mail.read", "notes.read.all", "mailboxsettings.readwrite", "Files.ReadWrite.All", "mail.send", "files.read", "files.read.all"]);
// The `trusted_apps` are sample App Client IDs. A MATCH will cause the alert to NOT fire; please modify for your environment!
let trusted_apps = dynamic([
'5523c0ac-586d-491e-bc73-fc18f9ed2c28', // DocuSign NA1
'3190d40e-82d9-44ee-af78-4ce2aafd4f15', // Zoom Video Communications
'0771c6f9-71c7-46d4-87a9-3b57443ee250', // Zoom OneDrive
'a644e37f-d354-440f-843a-1e1ee7dfc4ef', // Windows Defender Security Intelligence
'21fbe0e7-6e83-4fb0-a56f-92a67c64e37a', // Amazon Alexa Connect
'30966914-6789-4d1c-9e2b-76a81a5e2a54', // Barracuda Networks
'2a6082d8-9f0a-455c-a26b-7fce81edefb8', // Favro
'adc8d359-f3ef-4e0a-b170-2bcf5a022728' //ClassFlow Prod OneDrive Personal
]);
AuditLogs
| where TimeGenerated > ago(detectionTime)
| where LoggedByService =~ "Core Directory"
| where Category =~ "ApplicationManagement"
| where OperationName =~ "Consent to application"
| extend AppDisplayName = TargetResources.[0].displayName
| extend AppClientId = tolower(TargetResources.[0].id)
| where AppClientId !in (trusted_apps)
| extend ConsentFull = TargetResources[0].modifiedProperties[4].newValue
| parse ConsentFull with * "ConsentType: " GrantConsentType ", Scope: " GrantScope1 ", CreatedDateTime" * "]" *
| where GrantConsentType != "AllPrincipals" // NOTE: we are ignoring if OAuth application was granted to all users via an admin - but admin due diligence should be audited occasionally
| where ConsentFull has_any (o365_attack)
| extend GrantScopeCount = countof(tolower(GrantScope1), o365_attack_regex, 'regex')
| where GrantScopeCount > threshold
| extend GrantIpAddress = iff(isnotempty(InitiatedBy.user.ipAddress), tostring(InitiatedBy.user.ipAddress), tostring(InitiatedBy.app.ipAddress))
| extend GrantInitiatedBy = iff(isnotempty(InitiatedBy.user.userPrincipalName), tostring(InitiatedBy.user.userPrincipalName), tostring(InitiatedBy.app.displayName))
//
| extend GrantUserAgent = iff(AdditionalDetails[0].key =~ "User-Agent", tostring(AdditionalDetails[0].value), "")
| project TimeGenerated, GrantConsentType, GrantScope1, GrantInitiatedBy, AppDisplayName, GrantIpAddress, GrantUserAgent, AppClientId, OperationName, ConsentFull, CorrelationId
| join kind = leftouter (AuditLogs
| where TimeGenerated > ago(joinLookback)
| where LoggedByService =~ "Core Directory"
| where Category =~ "ApplicationManagement"
| where OperationName =~ "Add service principal"
| extend AppClientId = tolower(TargetResources[0].id)
| extend AppReplyURLs = iff(TargetResources[0].modifiedProperties[1].newValue has "AddressType", TargetResources[0].modifiedProperties[1].newValue, "")
| distinct AppClientId, tostring(AppReplyURLs)
)
on AppClientId
| join kind = innerunique (AuditLogs
| where TimeGenerated > ago(joinLookback)
| where LoggedByService =~ "Core Directory"
| where Category =~ "ApplicationManagement"
| where OperationName =~ "Add OAuth2PermissionGrant" or OperationName =~ "Add delegated permission grant"
| extend GrantAuthentication = tostring(TargetResources[0].displayName)
| extend GrantOperation = OperationName
| project GrantAuthentication, GrantOperation, CorrelationId
) on CorrelationId
| project TimeGenerated, GrantConsentType, GrantScope1, GrantInitiatedBy, AppDisplayName, AppReplyURLs, GrantIpAddress, GrantUserAgent, AppClientId, GrantAuthentication, OperationName, GrantOperation, CorrelationId, ConsentFull
| extend timestamp = TimeGenerated, AccountCustomEntity = GrantInitiatedBy, IPCustomEntity = GrantIpAddress
```

Comment (devikamehra) [reply Solutions/Azure Active Directory/Analytic Rules/MaliciousOAuthApp_O365AttackToolkit.yaml]: @treyperrone Thank you for working on this.
I will confirm on the trusted app list once again. You are right, the list should be dynamic without much changes to the query. Each approach has its own advantages and disadvantages.
For removing the filters from the join query, this might have performance impact when data is huge. Let's add those in to avoid any query timeouts.

Comment (devikamehra) [reply Solutions/Azure Active Directory/Analytic Rules/MaliciousOAuthApp_O365AttackToolkit.yaml]: @treyperrone 
1. Let's keep the source of trusted app based on the centralized csv for now. We can improve it in future, if required.
2. No, what I am saying is that we should keep the filters as-is in both part of the joins. There is no need to remove any filters from right side of the query. In case you are facing a performance issue, please let us know and we can take adequate steps.

Looking forward to your updated query.

File: Solutions/Azure Active Directory/Analytic Rules/MaliciousOAuthApp_O365AttackToolkit.yaml
Comment (devikamehra) [inline Solutions/Azure Active Directory/Analytic Rules/MaliciousOAuthApp_O365AttackToolkit.yaml:48]: My apologies for the confusion. The trusted apps were hardcoded for testing only. We should use the csv for now.

File: Solutions/Azure Active Directory/Analytic Rules/MaliciousOAuthApp_O365AttackToolkit.yaml
Comment (devikamehra) [inline Solutions/Azure Active Directory/Analytic Rules/MaliciousOAuthApp_O365AttackToolkit.yaml:49]: Yes, we can surely improve this using mv-apply with required DisplayName.

**Review by devikamehra — CHANGES_REQUESTED** [review]

File: Sample Data/Feeds/Microsoft.OAuth.KnownApplications.csv
Comment (devikamehra) [inline Sample Data/Feeds/Microsoft.OAuth.KnownApplications.csv:1]: Please do not delete these IDs. Kindly remove them from the code.

**Review by devikamehra — CHANGES_REQUESTED** [review]

File: Solutions/Azure Active Directory/Analytic Rules/MaliciousOAuthApp_O365AttackToolkit.yaml
Comment (treyperrone) [reply Solutions/Azure Active Directory/Analytic Rules/MaliciousOAuthApp_O365AttackToolkit.yaml:48]: I have feelings about this:

- If we are providing a default list it needs to be trackable as a change on the consumer side

- Since the list is completely separate consumers have no knowledge of the updates taking place to the list; it is not versioned with the rule. IF that list changes entirely next week it is completely invisible to consumer.

- How does a customer easily override some items? example; many government places ban Zoom app and use MS Teams only since it is better tied into AAD, SEntinel, Purview, etc.; consumer NOW has more work to do to override the list; it wont be simple for them to comment out and add. 

- They'd have to create their own dynamic list internally or publish an internal CSV file to consume and either way it presents more challenges to DIFF the rules on version updates.

**Comment by v-mchatla** [comment]
Hi @treyperrone - Can you please confirm once you have addressed all the queries so that Devika can start review on latest changes.
Thanks

File: Solutions/Azure Active Directory/Analytic Rules/MaliciousOAuthApp_O365AttackToolkit.yaml
Comment (devikamehra) [inline Solutions/Azure Active Directory/Analytic Rules/MaliciousOAuthApp_O365AttackToolkit.yaml:81]: Everything else looks good to me. Thank you for your patience. I will confirm with the author before the approval.
A small change, the changes made are modification in the logic. Thus version should be 1.1.0 now.

Comment (treyperrone) [reply Solutions/Azure Active Directory/Analytic Rules/MaliciousOAuthApp_O365AttackToolkit.yaml]: Screenshot 2023-03-08 at 6 41 55 AM

**Review by devikamehra — APPROVED** [review]
The author has also confirmed on the logic and approves the changes.

**Review by v-rucdu — APPROVED** [review]

**Pull request merged**
